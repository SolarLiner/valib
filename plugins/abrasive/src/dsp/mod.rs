use crate::{
    dsp::filter::{FilterModule, FilterParams},
    MAX_BLOCK_SIZE,
};
use nalgebra::Complex;
use nih_plug::prelude::*;
use std::sync::Arc;
use valib::Scalar;
use valib::{
    dsp::{analysis::DspAnalysis, blocks::Series, BlockAdapter},
    oversample::{Oversample, Oversampled},
};
use valib::{
    dsp::{DSPMeta, DSPProcess},
    simd::SimdComplexField,
};

pub mod filter;

pub const NUM_BANDS: usize = 5;

#[derive(Debug, Params)]
pub struct DspParams {
    #[id = "drive"]
    pub drive: FloatParam,
    #[id = "scale"]
    pub scale: FloatParam,
    #[nested(array)]
    pub filters: [Arc<FilterParams>; NUM_BANDS],
}

impl Default for DspParams {
    fn default() -> Self {
        Self {
            drive: FloatParam::new(
                "Drive",
                1.,
                FloatRange::Skewed {
                    min: 1e-2,
                    max: 100.,
                    factor: FloatRange::gain_skew_factor(-40., 40.),
                },
            )
            .with_string_to_value(formatters::s2v_f32_gain_to_db())
            .with_value_to_string(formatters::v2s_f32_gain_to_db(2))
            .with_unit("dB")
            .with_smoother(SmoothingStyle::Exponential(100.)),
            scale: FloatParam::new("Scale", 1., FloatRange::Linear { min: -1., max: 1. })
                .with_string_to_value(formatters::s2v_f32_percentage())
                .with_value_to_string(formatters::v2s_f32_percentage(2))
                .with_smoother(SmoothingStyle::Exponential(100.))
                .with_unit("%"),
            filters: std::array::from_fn(|_| Default::default()),
        }
    }
}

pub struct Equalizer<T> {
    params: Arc<DspParams>,
    dsp: Series<[FilterModule<T>; NUM_BANDS]>,
}

impl<T: Scalar> DSPMeta for Equalizer<T> {
    type Sample = T;
    fn set_samplerate(&mut self, samplerate: f32) {
        self.dsp.set_samplerate(samplerate);
    }

    fn latency(&self) -> usize {
        self.dsp.latency()
    }

    fn reset(&mut self) {
        self.dsp.reset()
    }
}

impl<T: Scalar> DSPProcess<1, 1> for Equalizer<T> {
    fn process(&mut self, [x]: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let drive = T::from_f64(self.params.drive.smoothed.next() as _);
        let scale = T::from_f64(self.params.scale.smoothed.next() as _);
        for filter in &mut self.dsp.0 {
            filter.set_scale(scale);
        }
        let [y] = self.dsp.process([drive * x]);
        [y / drive]
    }
}

impl<T: Scalar> DspAnalysis<1, 1> for Equalizer<T> {
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; 1]; 1] {
        self.dsp.h_z(z)
    }

    fn freq_response(
        &self,
        samplerate: Self::Sample,
        f: Self::Sample,
    ) -> [[Complex<Self::Sample>; 1]; 1]
    where
        Complex<Self::Sample>: SimdComplexField,
    {
        self.dsp.freq_response(samplerate, f)
    }
}

impl<T: Scalar> Equalizer<T> {
    pub fn new(samplerate: T, params: Arc<DspParams>) -> Self {
        let dsp = Series(std::array::from_fn(|i| {
            FilterModule::new(samplerate, params.filters[i].clone())
        }));
        Self { params, dsp }
    }

    pub fn use_param_values(&mut self, use_modulated: bool) {
        for filter in &mut self.dsp.0 {
            filter.use_param_values(use_modulated);
        }
    }
}

pub const OVERSAMPLE: usize = 4;

pub type Dsp<T> = Oversampled<T, BlockAdapter<Equalizer<T>>>;
pub fn create<T: Scalar>(samplerate: f32, params: Arc<DspParams>) -> Dsp<T>
where
    Complex<T>: SimdComplexField,
{
    let target_samplerate = OVERSAMPLE as f32 * samplerate;
    Oversample::new(OVERSAMPLE, MAX_BLOCK_SIZE).with_dsp(
        samplerate,
        BlockAdapter(Equalizer::new(T::from_f64(target_samplerate as _), params)),
    )
}
