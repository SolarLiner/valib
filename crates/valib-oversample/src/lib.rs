#![warn(missing_docs)]
//! # Oversampling
//!
//! This crate provides oversampling capabilities, allowing you to oversample any block processor.
//!
//! Available here is a polyphase-based oversampling method, with more to come in the future.

use num_complex::Complex;

use valib_core::dsp::buffer::{AudioBufferMut, AudioBufferRef};
use valib_core::dsp::parameter::HasParameters;
use valib_core::dsp::DSPProcessBlock;
use valib_core::dsp::{DSPMeta, DSPProcess};
use valib_core::simd::SimdComplexField;
use valib_core::Scalar;
use valib_filters::halfband;
use valib_filters::halfband::HalfbandFilter;

/// Ping-pong buffer. Allows processing of effect chains operating on buffers, by allowing the input
/// and output buffers be swapped after each effect.
#[derive(Debug, Clone)]
pub struct PingPongBuffer<T> {
    left: Box<[T]>,
    right: Box<[T]>,
    input_is_left: bool,
}

impl<T> PingPongBuffer<T> {
    /// Create a new ping-pong buffer
    ///
    /// # Arguments
    ///
    /// * `contents`: Initial contents of the buffers
    ///
    /// returns: PingPongBuffer<T>
    pub fn new<I: IntoIterator<Item = T>>(contents: I) -> Self
    where
        I::IntoIter: Clone,
    {
        let it = contents.into_iter();
        Self {
            left: it.clone().collect(),
            right: it.collect(),
            input_is_left: true,
        }
    }

    /// Fill the buffers
    ///
    /// # Arguments
    ///
    /// * `value`:
    ///
    /// returns: ()
    ///
    /// # Examples
    ///
    /// ```
    ///
    /// ```
    pub fn fill(&mut self, value: T)
    where
        T: Copy,
    {
        self.left.fill(value);
        self.right.fill(value);
    }

    /// Get the input and output buffers.
    ///
    /// # Arguments
    ///
    /// * `index`: Value to slice the indices with, or .. to get the entire buffer at once.
    ///
    /// returns: (&[T], &mut [T])
    pub fn get_io_buffers<I: Clone>(&mut self, index: I) -> (&[T], &mut [T])
    where
        [T]: std::ops::IndexMut<I, Output = [T]>,
    {
        if self.input_is_left {
            let input = &self.left[index.clone()];
            let output = &mut self.right[index];
            (input, output)
        } else {
            let input = &self.right[index.clone()];
            let output = &mut self.left[index];
            (input, output)
        }
    }

    /// Get an immutable reference to the output buffer
    ///
    /// # Arguments
    ///
    /// * `index`: Index value to slice the buffer with, or .. to get the entire buffer at once
    ///
    /// returns: &[T]
    pub fn get_output_ref<I>(&self, index: I) -> &[T]
    where
        [T]: std::ops::Index<I, Output = [T]>,
    {
        if self.input_is_left {
            &self.right[index]
        } else {
            &self.left[index]
        }
    }

    /// Copy the output buffer of this ping-pong buffer into the provided output buffer.
    ///
    /// # Arguments
    ///
    /// * `output`: Output buffer into which the inner output buffer will be copied into.
    ///
    /// returns: ()
    pub fn copy_into(&self, output: &mut [T])
    where
        T: Copy,
    {
        let slice = if self.input_is_left {
            &self.right[..output.len()]
        } else {
            &self.left[..output.len()]
        };
        output.copy_from_slice(slice);
    }

    /// Switch the buffers around.
    pub fn switch(&mut self) {
        self.input_is_left = !self.input_is_left;
    }

    /// Returns true if the ping-pong buffers are empty.
    #[allow(dead_code)]
    pub fn is_empty(&self) -> bool {
        self.left.is_empty()
    }

    /// Returns the ping-pong buffer length.
    pub fn len(&self) -> usize {
        self.left.len()
    }
}

/// Single resample stage.
#[derive(Debug, Clone, Copy)]
pub struct ResampleStage<T, const UPSAMPLE: bool> {
    filter: HalfbandFilter<T, 6>,
}

impl<T: Scalar, const UPSAMPLE: bool> Default for ResampleStage<T, UPSAMPLE> {
    fn default() -> Self {
        Self {
            filter: halfband::steep_order12(),
        }
    }
}

impl<T: Scalar, const UPSAMPLE: bool> ResampleStage<T, UPSAMPLE> {
    /// Latency of the resample stage
    pub fn latency(&self) -> usize {
        self.filter.latency()
    }

    /// Reset the resample stage
    pub fn reset(&mut self) {
        self.filter.reset();
    }
}

impl<T: Scalar> ResampleStage<T, true> {
    /// Upsample a single sample of audio
    ///
    /// # Arguments
    ///
    /// * `s`: Input sample
    ///
    /// returns: [T; 2]
    pub fn process(&mut self, s: T) -> [T; 2] {
        let [x0] = self.filter.process([s + s]);
        let [x1] = self.filter.process([T::zero()]);
        [x0, x1]
    }

    /// Upsample the input buffer by a factor of 2.
    ///
    /// The output slice should be twice the length of the input slice.
    #[allow(clippy::identity_op)]
    pub fn process_block(&mut self, input: &[T], output: &mut [T]) {
        assert_eq!(input.len() * 2, output.len());
        for (i, s) in input.iter().copied().enumerate() {
            let [x0, x1] = self.process(s);
            output[2 * i + 0] = x0;
            output[2 * i + 1] = x1;
        }
    }
}

impl<T: Scalar> ResampleStage<T, false> {
    /// Downsample 2 samples of input audio, and output a single sample of audio.
    ///
    /// # Arguments
    ///
    /// * `[x0, x1]`: Inputs samples
    ///
    /// returns: T
    pub fn process(&mut self, [x0, x1]: [T; 2]) -> T {
        let [y] = self.filter.process([x0]);
        let _ = self.filter.process([x1]);
        y
    }

    /// Downsample the input buffer by a factor of 2.
    ///
    /// The output slice should be twice the length of the input slice.
    #[allow(clippy::identity_op)]
    pub fn process_block(&mut self, input: &[T], output: &mut [T]) {
        assert_eq!(input.len(), 2 * output.len());
        for i in 0..output.len() {
            let [y] = self.filter.process([input[2 * i + 0]]);
            let [_] = self.filter.process([input[2 * i + 1]]);
            output[i] = y;
        }
    }
}

/// Raw oversampling type. Works by taking a block of audio, processing it and returning a slice to
/// an internal buffer containing the upsampled audio data you should process in place. Once done,
/// call `.finish(output)` on the slice to downsample the internal buffer again, and output it to
/// `output`.
#[derive(Debug, Clone)]
pub struct Oversample<T> {
    max_factor: usize,
    num_stages_active: usize,
    os_buffer: PingPongBuffer<T>,
    upsample: Box<[ResampleStage<T, true>]>,
    downsample: Box<[ResampleStage<T, false>]>,
}

impl<T> Oversample<T> {
    /// Returns the current oversampling amount.
    pub fn oversampling_amount(&self) -> usize {
        usize::pow(2, self.num_stages_active as _)
    }

    /// Sets the oversampling amount.
    ///
    /// Only square numbers are supported; otherwise the next power of two from the given factor
    /// will be used.
    ///
    /// # Arguments
    ///
    /// `amt`: Oversampling amount. Needs to be less than or equal to the maximum oversampling rate
    ///     configured when constructed with [`Oversample::new`].
    pub fn set_oversampling_amount(&mut self, amt: usize) {
        assert!(amt <= self.max_factor);
        self.num_stages_active = amt.next_power_of_two().ilog2() as _;
    }

    /// Maximum block size supported at the current oversampling factor.
    pub fn max_block_size(&self) -> usize {
        self.os_buffer.len() / usize::pow(2, self.num_stages_active as _)
    }

    /// Return the length of the oversampled buffer.
    pub fn get_os_len(&self, input_len: usize) -> usize {
        input_len * usize::pow(2, self.num_stages_active as _)
    }
}

impl<T: Scalar> Oversample<T> {
    /// Create a new oversampling filter.
    ///
    /// # Arguments
    ///
    /// * `max_os_factor`: Maximum oversampling factor supported by this instance. The actual
    ///     oversampling can be changed after creation, but will need to always be less than or equal
    ///     to this factor.
    /// * `max_block_size`: Maximum block size that will be expected to be processed.
    ///
    /// returns: Oversample<T>
    pub fn new(max_os_factor: usize, max_block_size: usize) -> Self
    where
        Complex<T>: SimdComplexField,
    {
        assert!(max_os_factor >= 1);
        let max_os_factor = max_os_factor.next_power_of_two();
        let num_stages = max_os_factor.ilog2() as usize;
        let os_buffer = vec![T::zero(); max_block_size * max_os_factor];
        let os_buffer = PingPongBuffer::new(os_buffer);
        let upsample = (0..num_stages).map(|_| ResampleStage::default()).collect();
        let downsample = (0..num_stages).map(|_| ResampleStage::default()).collect();
        Self {
            max_factor: max_os_factor,
            num_stages_active: num_stages,
            os_buffer,
            upsample,
            downsample,
        }
    }

    /// Returns the latency of the filter. This includes both upsampling and downsampling.
    pub fn latency(&self) -> usize {
        let upsample_latency = self.upsample.iter().map(|p| p.latency()).sum::<usize>();
        let downsample_latency = self.downsample.iter().map(|p| p.latency()).sum::<usize>();
        2 * self.num_stages_active + upsample_latency + downsample_latency
    }

    /// Reset the state of this oversampling filter.
    pub fn reset(&mut self) {
        self.os_buffer.fill(T::zero());
        for stage in &mut self.upsample {
            stage.reset();
        }
        for stage in &mut self.downsample {
            stage.reset();
        }
    }

    /// Construct an [`Oversampled`] given this oversample instance and a block processor to wrap.
    pub fn with_dsp<P: DSPProcessBlock<1, 1>>(
        self,
        samplerate: f32,
        mut dsp: P,
    ) -> Oversampled<T, P> {
        let max_block_size = dsp.max_block_size().unwrap_or(self.os_buffer.len());
        // Verify that we satisfy the inner DSPBlock instance's requirement on maximum block size
        assert!(self.os_buffer.len() <= max_block_size);
        let staging_buffer = vec![T::zero(); max_block_size].into_boxed_slice();
        dsp.set_samplerate(samplerate * self.num_stages_active as f32);
        Oversampled {
            oversampling: self,
            staging_buffer,
            inner: dsp,
            base_samplerate: samplerate,
        }
    }

    #[profiling::function]
    fn upsample(&mut self, input: &[T]) -> &mut [T] {
        assert!(input.len() <= self.max_block_size());
        if self.num_stages_active == 0 {
            let (_, output) = self.os_buffer.get_io_buffers(..input.len());
            output.copy_from_slice(input);
            return output;
        }

        let os_len = self.get_os_len(input.len());
        let mut len = input.len();
        let (_, output) = self.os_buffer.get_io_buffers(..len);
        output.copy_from_slice(input);
        for stage in &mut self.upsample[..self.num_stages_active] {
            self.os_buffer.switch();
            let (input, output) = self.os_buffer.get_io_buffers(..2 * len);
            stage.process_block(&input[..len], output);
            len *= 2;
        }
        let (_, output) = self.os_buffer.get_io_buffers(..os_len);
        output
    }

    #[profiling::function]
    fn downsample(&mut self, out: &mut [T]) {
        if self.num_stages_active == 0 {
            let inner_out = self.os_buffer.get_output_ref(..out.len());
            out.copy_from_slice(inner_out);
            return;
        }

        let os_len = self.get_os_len(out.len());
        let mut len = os_len;
        for stage in &mut self.downsample {
            self.os_buffer.switch();
            let (input, output) = self.os_buffer.get_io_buffers(..len);
            stage.process_block(input, &mut output[..len / 2]);
            len /= 2;
        }
        self.os_buffer.copy_into(out);
    }
}

/// Wraps a block processor to orversample it, and allow using it within other DSP blocks.
///
/// Oversampling is transparently performed over the inner block processor.
pub struct Oversampled<T, P> {
    oversampling: Oversample<T>,
    staging_buffer: Box<[T]>,
    /// Inner processor
    pub inner: P,
    base_samplerate: f32,
}

impl<T, P> Oversampled<T, P> {
    /// Return the current oversampling factor
    pub fn os_factor(&self) -> usize {
        self.oversampling.oversampling_amount()
    }

    /// Drops the oversampling filter, returning the inner processor.
    pub fn into_inner(self) -> P {
        self.inner
    }
}

impl<T, P> Oversampled<T, P>
where
    T: Scalar,
    P: DSPProcessBlock<1, 1, Sample = T>,
{
    /// Sets the oversampling amount. See [`Oversample::set_oversampling_amount`] for more details.
    pub fn set_oversampling_amount(&mut self, amt: usize) {
        assert!(amt >= 1);
        self.oversampling.set_oversampling_amount(amt);
        self.set_samplerate(self.base_samplerate);
    }

    /// Returns the sample rate of the oversampled buffer.
    pub fn inner_samplerate(&self) -> f32 {
        self.base_samplerate * self.oversampling.oversampling_amount() as f32
    }
}

impl<T: Scalar, P: DSPMeta<Sample = T>> DSPMeta for Oversampled<T, P> {
    type Sample = T;

    fn set_samplerate(&mut self, samplerate: f32) {
        self.inner
            .set_samplerate(self.os_factor() as f32 * samplerate);
    }

    fn latency(&self) -> usize {
        self.inner.latency()
    }

    fn reset(&mut self) {
        self.oversampling.reset();
        self.inner.reset();
    }
}

#[profiling::all_functions]
impl<T, P> DSPProcessBlock<1, 1> for Oversampled<T, P>
where
    Self: DSPMeta<Sample = T>,
    T: Scalar,
    P: DSPProcessBlock<1, 1, Sample = T>,
{
    fn process_block(&mut self, inputs: AudioBufferRef<T, 1>, mut outputs: AudioBufferMut<T, 1>) {
        let os_block = self.oversampling.upsample(inputs.get_channel(0));

        let mut inner_input =
            AudioBufferMut::new([&mut self.staging_buffer[..os_block.len()]]).unwrap();
        inner_input.copy_from_slice(0, os_block);
        let inner_output = AudioBufferMut::new([os_block]).unwrap();

        self.inner.process_block(inner_input.as_ref(), inner_output);

        self.oversampling.downsample(outputs.get_channel_mut(0));
    }

    fn max_block_size(&self) -> Option<usize> {
        Some(self.oversampling.max_block_size())
    }
}

impl<S, P: HasParameters> HasParameters for Oversampled<S, P> {
    type Name = P::Name;

    fn set_parameter(&mut self, param: Self::Name, value: f32) {
        self.inner.set_parameter(param, value)
    }
}

#[cfg(test)]
mod tests {
    use numeric_literals::replace_float_literals;
    use valib_core::dsp::{buffer::AudioBufferBox, DSPProcess, DSPProcessBlock as _};
    use valib_core::Scalar;
    use valib_core::{
        dsp::{BlockAdapter, DSPMeta},
        util::tests::{Plot, Series},
    };

    use super::{Oversample, PingPongBuffer};

    #[test]
    fn ping_pong_works() {
        let mut pingpong = PingPongBuffer::new([0; 8]);
        let (input, output) = pingpong.get_io_buffers(..);
        assert_eq!(0, input[0]);
        assert_eq!(0, output[0]);

        output[0] = 1;
        pingpong.switch();

        let (input, output) = pingpong.get_io_buffers(..);
        assert_eq!(1, input[0]);
        assert_eq!(0, output[0]);
    }

    #[test]
    fn oversampled_dsp_block() {
        use plotters::prelude::*;

        struct NaiveSquare<T> {
            samplerate: T,
            frequency: T,
            phase: T,
        }
        impl<T: Scalar> DSPMeta for NaiveSquare<T> {
            type Sample = T;
        }

        impl<T: Scalar> DSPProcess<1, 1> for NaiveSquare<T> {
            #[replace_float_literals(T::from_f64(literal))]
            fn process(&mut self, _: [Self::Sample; 1]) -> [Self::Sample; 1] {
                let step = self.frequency / self.samplerate;
                let out = (1.0).select(self.phase.simd_gt(0.5), -1.0);
                self.phase += step;
                self.phase = (self.phase - 1.0).select(self.phase.simd_gt(1.0), self.phase);
                [out]
            }
        }

        let samplerate = 1000f32;
        let freq = 10f32;
        let dsp = NaiveSquare {
            samplerate: 4.0 * samplerate,
            frequency: freq,
            phase: 0.0,
        };
        let mut os = Oversample::<f32>::new(4, 64).with_dsp(samplerate, BlockAdapter(dsp));

        let input = AudioBufferBox::zeroed(64);
        let mut output = AudioBufferBox::zeroed(64);
        os.process_block(input.as_ref(), output.as_mut());
        Plot {
            title: "Oversample block",
            bode: false,
            series: &[
                Series {
                    label: "Oversampled",
                    samplerate: 4. * samplerate,
                    series: os.oversampling.os_buffer.get_output_ref(..),
                    color: &YELLOW,
                },
                Series {
                    label: "Output",
                    samplerate,
                    series: output.get_channel(0),
                    color: &BLUE,
                },
            ],
        }
        .create_svg("plots/oversample/dsp_block.svg");
        insta::assert_csv_snapshot!(output.get_channel(0), { "[]" => insta::rounded_redaction(3) });
    }
}
