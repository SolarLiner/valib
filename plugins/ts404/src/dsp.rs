use valib::dsp::DSP;
use valib::filters::statespace::StateSpace;
use valib::saturators::Slew;
use valib::Scalar;

#[derive(Debug, Copy, Clone)]
pub struct ClipperStage<T: Scalar>(StateSpace<T, 1, 3, 1>, Slew<T>);

impl<T: Scalar> DSP<1, 1> for ClipperStage<T> {
    type Sample = T;

    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let [y] = self.0.process(x);
        self.1.process([y.simd_asinh()])
    }

    fn latency(&self) -> usize {
        self.0.latency() + self.1.latency()
    }

    fn reset(&mut self) {
        self.0.reset();
        self.1.reset();
    }
}

impl<T: Scalar> ClipperStage<T> {
    pub fn new(samplerate: T, dist: T) -> Self {
        let dt = samplerate.simd_recip();
        Self(crate::gen::clipper(dist, dt), Slew::new(T::from_f64(1e5) * dt))
    }

    pub fn set_params(&mut self, samplerate: T, dist: T) {
        let dt = samplerate.simd_recip();
        self.0.update_matrices(&crate::gen::clipper(dist, dt));
        self.1.set_max_diff(T::from_f64(1e5), dt);
    }
}

#[derive(Debug, Copy, Clone)]
pub struct ToneStage<T: Scalar>(StateSpace<T, 1, 4, 1>);

impl<T: Scalar> DSP<1, 1> for ToneStage<T> {
    type Sample = T;

    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        self.0.process(x)
    }

    fn latency(&self) -> usize {
        self.0.latency()
    }

    fn reset(&mut self) {
        self.0.reset();
    }
}

impl<T: Scalar> ToneStage<T> {
    pub fn new(samplerate: T, tone: T) -> Self {
        Self(crate::gen::tone(tone, samplerate.simd_recip()))
    }

    pub fn update_params(&mut self, samplerate: T, tone: T) {
        self.0.update_matrices(&crate::gen::tone(tone, samplerate.simd_recip()));
    }
}