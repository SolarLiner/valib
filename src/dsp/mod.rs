#![doc = include_str!("./README.md")]

use nalgebra::Complex;
use num_traits::Zero;

use crate::dsp::buffer::{AudioBufferBox, AudioBufferMut, AudioBufferRef};
use crate::Scalar;

use self::analysis::DspAnalysis;

pub mod analysis;
pub mod blocks;
pub mod buffer;
pub mod parameter;
pub mod utils;

/// Trait for interacting with a DSP algorithm, outside of processing. Shared by processors of both
/// per-sample algorithms and block-based algorithms.
#[allow(unused_variables)]
pub trait DSPMeta {
    /// Type of the audio sample used by this DSP instance.
    type Sample: Scalar;

    /// Sets the processing samplerate for this [`DSPProcess`] instance.
    fn set_samplerate(&mut self, samplerate: f32) {}

    /// Report the latency of this DSP instance, that is the time, in samples, it takes for an input sample to be
    /// output back.
    fn latency(&self) -> usize {
        0
    }

    /// Reset this instance. Parameters should be kept, but any memory and derived state should be put back to a
    /// well-known default value.
    fn reset(&mut self) {}
}

/// DSP trait. This is the main abstraction of the whole library.
///
/// Implementors of this trait are processes that work on a per-sample basis.
///
/// Multichannel I/O is supported, and is determined by the `I` and `O` const generics.
/// It's up to each implementor to document what the inputs and outputs mean.
/// There are no restrictions on the number of input or output channels in and of itself.
/// Implementors can support multiple configurations of I/O for different setups.
pub trait DSPProcess<const I: usize, const O: usize>: DSPMeta {
    /// Process a single sample of audio. No assumptions are made on the contents of said sample,
    /// so it can both work for working with audio data, but also control signals like frequency in Hertz
    /// for oscillators, or gate signals that are actually either 0 or 1.
    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O];
}

/// Trait for DSP processes that take in buffers of audio instead of single-samples.
/// Documentation of [`DSPProcess`] still applies in here; only the process method changes.
pub trait DSPProcessBlock<const I: usize, const O: usize>: DSPMeta {
    /// Process a block of audio. Implementors should take the values of inputs and produce a stream into outputs,
    /// *as if* it was processed sample-by-sample. Ideally there should be no difference between a [`DSPProcess`] and [`DSPProcessBlock`]
    /// implementation, when both can exist. Note that [`DSPProcessBlock`] is blanket-implemented for all [`DSPProcess`] implementors,
    /// and as such, a simultaneous implementation of both onto one struct is impossible.
    ///
    /// Implementors should assume inputs and outputs are of the same length, as it is the caller's responsibility to make sure of that.
    fn process_block(
        &mut self,
        inputs: AudioBufferRef<Self::Sample, I>,
        outputs: AudioBufferMut<Self::Sample, O>,
    );

    /// Define an optional maximum buffer size alloed by this [`DSPProcessBlock`] instance. Callers into this instance must
    /// then only provide buffers that are up to this size in samples.
    #[inline(always)]
    fn max_block_size(&self) -> Option<usize> {
        None
    }
}

impl<P, const I: usize, const O: usize> DSPProcessBlock<I, O> for P
where
    P: DSPProcess<I, O>,
{
    #[inline(never)]
    fn process_block(
        &mut self,
        inputs: AudioBufferRef<P::Sample, I>,
        mut outputs: AudioBufferMut<P::Sample, O>,
    ) {
        if I == 0 && O == 0 {
            return;
        }
        for i in 0..inputs.samples() {
            outputs.set_frame(i, self.process(inputs.get_frame(i)))
        }
    }
}

/// Adapt a [`DSPProcessBlock`] instance to be able to used as a [`DSPProcess`].
///
/// This introduces as much latency as the internal buffer size is.
/// The internal buffer size is determined by either the max accepted buffer size of the inner instance, or is set
/// to 64 samples by default.
pub struct PerSampleBlockAdapter<P, const I: usize, const O: usize>
where
    P: DSPProcessBlock<I, O>,
{
    input_buffer: AudioBufferBox<P::Sample, I>,
    input_filled: usize,
    output_buffer: AudioBufferBox<P::Sample, O>,
    output_filled: usize,
    inner: P,
    pub buffer_size: usize,
}

impl<P, const I: usize, const O: usize> std::ops::Deref for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    type Target = P;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<P, const I: usize, const O: usize> std::ops::DerefMut for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl<P, const I: usize, const O: usize> PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    pub const DEFAULT_BUFFER_SIZE: usize = 64;
    pub fn new(dsp_block: P) -> Self {
        Self::new_with_max_buffer_size(dsp_block, Self::DEFAULT_BUFFER_SIZE)
    }

    pub fn new_with_max_buffer_size(dsp_block: P, max_buffer_size: usize) -> Self {
        let buffer_size = dsp_block
            .max_block_size()
            .map(|mbs| mbs.min(max_buffer_size))
            .unwrap_or(max_buffer_size);
        Self {
            input_buffer: AudioBufferBox::zeroed(buffer_size),
            input_filled: 0,
            output_buffer: AudioBufferBox::zeroed(buffer_size),
            output_filled: buffer_size,
            buffer_size,
            inner: dsp_block,
        }
    }

    pub fn into_inner(self) -> P {
        self.inner
    }
}

impl<P, const I: usize, const O: usize> DSPMeta for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    type Sample = P::Sample;

    fn set_samplerate(&mut self, samplerate: f32) {
        self.inner.set_samplerate(samplerate);
    }

    fn latency(&self) -> usize {
        (self.inner.latency() + self.input_buffer.samples()).saturating_sub(1)
    }

    fn reset(&mut self) {
        self.input_filled = 0;
        self.output_filled = self.output_buffer.samples();
        self.input_buffer.fill(P::Sample::zero());
        self.output_buffer.fill(P::Sample::zero());
    }
}

impl<P, const I: usize, const O: usize> DSPProcess<I, O> for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O] {
        self.input_buffer.set_frame(self.input_filled, x);
        self.input_filled += 1;
        if self.input_buffer.samples() == self.input_filled {
            self.inner
                .process_block(self.input_buffer.as_ref(), self.output_buffer.as_mut());
            self.input_filled = 0;
            self.output_filled = 0;
        }

        if self.output_filled < self.buffer_size {
            let ret = self.output_buffer.get_frame(self.output_filled);
            self.output_filled += 1;
            ret
        } else {
            [Self::Sample::zero(); O]
        }
    }
}

impl<P, const I: usize, const O: usize> DspAnalysis<I, O> for PerSampleBlockAdapter<P, I, O>
where
    P: DspAnalysis<I, O>,
{
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; O]; I] {
        self.inner.h_z(z)
    }
}

#[cfg(test)]
mod tests {
    use std::marker::PhantomData;

    use super::*;

    #[test]
    fn test_per_sample_block_adapter() {
        struct Counter<T>(PhantomData<T>);

        impl<T: Scalar> Counter<T> {
            pub fn new() -> Self {
                Self(PhantomData)
            }
        }

        impl<T: Scalar> DSPMeta for Counter<T> {
            type Sample = T;
        }

        impl<T: Scalar> DSPProcessBlock<0, 1> for Counter<T> {
            fn process_block(
                &mut self,
                inputs: AudioBufferRef<T, 0>,
                mut outputs: AudioBufferMut<T, 1>,
            ) {
                let len = inputs.samples();
                assert_eq!(len, outputs.samples());

                for (i, out) in outputs.get_channel_mut(0).iter_mut().enumerate() {
                    *out = T::from_f64(i as _);
                }
            }
        }

        let mut adaptor = PerSampleBlockAdapter::new_with_max_buffer_size(Counter::<f32>::new(), 4);
        assert_eq!(3, adaptor.latency());

        let expected = [0.0, 0.0, 0.0, 0.0, 1.0, 2.0, 3.0, 0.0];
        let mut actual = [0.0; 8];

        // Calling `process_block` but it's actually calling the impl for all `DSP` passing each sample through.
        adaptor.process_block(
            AudioBufferRef::empty(8),
            AudioBufferMut::new([&mut actual]).unwrap(),
        );

        assert_eq!(expected, actual);
    }
}
