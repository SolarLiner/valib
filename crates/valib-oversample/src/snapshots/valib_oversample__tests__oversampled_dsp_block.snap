---
source: crates/valib-oversample/src/lib.rs
expression: output.get_channel(0)
---
-0.0
-0.002
-0.064
-0.44
-1.091
-1.132
-0.859
-1.083
-0.977
-0.98
-1.044
-0.947
-1.053
-0.952
-1.04
-0.97
-1.021
-0.987
-1.005
-1.001
-0.994
-1.011
-0.986
-1.016
-0.982
-1.019
-0.98
-1.02
-0.98
-1.019
-0.981
-1.018
-0.983
-1.016
-0.985
-1.014
-0.987
-1.012
-0.989
-1.01
-0.991
-1.008
-0.993
-1.006
-0.994
-1.005
-0.996
-1.003
-0.997
-1.002
-0.998
-1.0
-0.934
-0.397
0.896
1.405
0.754
1.051
1.073
0.873
1.136
0.881
1.091
0.94
